application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java

```diff
@@ -139,26 +138,58 @@ private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
-                actionsStore.getActionsByTargetAscending(guild.getIdLong(), user.getIdLong());
+    private Message auditMessage(long guildId, long targetId, int pageNo, @NotNull JDA jda) {
+        List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
 
-        MessageEmbed summary = createSummaryMessage(user, actions);
-        if (actions.isEmpty()) {
-            event.replyEmbeds(summary).queue();
-            return;
+        List<List<ActionRecord>> groupedActions = new ArrayList<>();
+        for (int i = 0; i < actions.size(); i++) {
+            if (i % 25 == 0) {
+                groupedActions.add(new ArrayList<>(25));
+            }
+
+            groupedActions.get(groupedActions.size() - 1).add(actions.get(i));
+        }
+
+        int totalPage = groupedActions.size();
+
+        EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions)
+            .setFooter("Page: " + pageNo + "/" + totalPage);
+        groupedActions.get(pageNo - 1)
+            .forEach(action -> audit.addField(actionToField(action, jda)));
+
+        return new MessageBuilder(audit.build())
+            .setActionRows(ActionRow.of(previousButton(guildId, targetId, pageNo),
+                    nextButton(guildId, targetId, pageNo, totalPage)))
+            .build();
+    }
+
+    private Button previousButton(long guildId, long targetId, int pageNo) {
+        Button previousButton = Button.primary(generateComponentId(String.valueOf(guildId),
+                String.valueOf(targetId), String.valueOf(pageNo), "-1"), "⬅");
+
+        if (pageNo == 1) {
+            previousButton = previousButton.asDisabled();
+        }
+
+        return previousButton;
+    }
+
+    private Button nextButton(long guildId, long targetId, int pageNo, int totalPage) {
+        Button nextButton = Button.primary(generateComponentId(String.valueOf(guildId),
+                String.valueOf(targetId), String.valueOf(pageNo), "1"), "➡");
+
+        if (pageNo == totalPage) {
+            nextButton = nextButton.asDisabled();
         }
 
-        // Computing messages for actual actions is done deferred and might require asking the
-        // Discord API
-        event.deferReply().queue();
-        JDA jda = event.getJDA();
+        return nextButton;
+    }
 
-        RestAction<List<MessageEmbed>> messagesTask = RestAction
-            .allOf(actions.stream().map(action -> actionToMessage(action, jda)).toList());
-        messagesTask.map(messages -> prependElement(summary, messages))
-            .flatMap(messages -> event.getHook().sendMessageEmbeds(messages))
+    @Override
+    public void onButtonClick(@NotNull ButtonInteractionEvent event, @NotNull List<String> args) {
+        event
+            .editMessage(auditMessage(Long.parseLong(args.get(0)), Long.parseLong(args.get(1)),
+                    Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3)), event.getJDA()))
```

Comment:
Looking at the `onButtonClick` method, it is not understandable what these args are for. Please put them into well named variables beforehand instead:
```java
Foo foo = args.get(0);
long bar = Long.parseLong(args.get(1));
int baz = Integer.parseInt(args.get(2));
int buz = Integer.parseInt(args.get(3));

event.editMessage(auditMessage(foo, bar, baz, buz, ...);
```
(but with proper names)

Reply:
You solved this only partially. It is still unclear, looking at just the code, what the role of args 3 and args 4 is.

Reply:
adding comment for this will do the thing?

Reply:
Or you just put it into variables. Code should be self-explanatory.
```java
int currentPage = ...
int turnPageBy = ...

int pageToDisplay = currentPage + turnPageBy;
```

Dont be scared of writing long code. If it actually helps readability, it is always worth it.